File: app/models/recipe.py
"""
菜谱相关数据库模型
"""
from sqlalchemy import Column, Integer, String, Float, Text, ForeignKey, JSON
from sqlalchemy.orm import relationship

from app.core.database import Base


class Recipe(Base):
    """菜谱表"""
    __tablename__ = "recipes"

    id = Column(String, primary_key=True, index=True)
    name = Column(String, nullable=False, index=True)
    description = Column(Text)
    source_path = Column(String)
    image_path = Column(String, nullable=True)
    images = Column(JSON, default=list)  # 存储图片数组
    category = Column(String, index=True)
    difficulty = Column(Integer)
    tags = Column(JSON, default=list)  # 存储标签数组
    servings = Column(Integer)
    prep_time_minutes = Column(Integer, nullable=True)
    cook_time_minutes = Column(Integer, nullable=True)
    total_time_minutes = Column(Integer, nullable=True)

    # 关联关系
    ingredients = relationship(
        "Ingredient",
        back_populates="recipe",
        cascade="all, delete-orphan",
        lazy="joined"
    )
    steps = relationship(
        "Step",
        back_populates="recipe",
        cascade="all, delete-orphan",
        order_by="Step.step",
        lazy="joined"
    )
    additional_notes = relationship(
        "AdditionalNote",
        back_populates="recipe",
        cascade="all, delete-orphan",
        lazy="joined"
    )


class Ingredient(Base):
    """食材表"""
    __tablename__ = "ingredients"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    recipe_id = Column(String, ForeignKey("recipes.id", ondelete="CASCADE"), nullable=False, index=True)
    name = Column(String, nullable=False)
    quantity = Column(Float, nullable=True)
    unit = Column(String, nullable=True)
    text_quantity = Column(String, nullable=False)
    notes = Column(String, nullable=True)

    # 关联关系
    recipe = relationship("Recipe", back_populates="ingredients")


class Step(Base):
    """步骤表"""
    __tablename__ = "steps"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    recipe_id = Column(String, ForeignKey("recipes.id", ondelete="CASCADE"), nullable=False, index=True)
    step = Column(Integer, nullable=False)  # 步骤序号
    description = Column(Text, nullable=False)

    # 关联关系
    recipe = relationship("Recipe", back_populates="steps")


class AdditionalNote(Base):
    """小贴士表"""
    __tablename__ = "additional_notes"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    recipe_id = Column(String, ForeignKey("recipes.id", ondelete="CASCADE"), nullable=False, index=True)
    note = Column(Text, nullable=False)

    # 关联关系
    recipe = relationship("Recipe", back_populates="additional_notes")

File: internal/models/recipe.go
package models

import (
	"database/sql/driver"
	"encoding/json"
	"errors"
)

// StringSlice JSON 存储的字符串数组
type StringSlice []string

func (s StringSlice) Value() (driver.Value, error) {
	if s == nil {
		return "[]", nil
	}
	return json.Marshal(s)
}

func (s *StringSlice) Scan(value interface{}) error {
	if value == nil {
		*s = []string{}
		return nil
	}

	var bytes []byte
	switch v := value.(type) {
	case []byte:
		bytes = v
	case string:
		bytes = []byte(v)
	default:
		return errors.New("无法将值转换为 StringSlice")
	}

	return json.Unmarshal(bytes, s)
}

// Recipe 菜谱表
type Recipe struct {
	ID               string      `gorm:"primaryKey;column:id" json:"id"`
	Name             string      `gorm:"not null;index" json:"name"`
	Description      *string     `gorm:"type:text" json:"description"`
	SourcePath       *string     `gorm:"column:source_path" json:"source_path"`
	ImagePath        *string     `gorm:"column:image_path" json:"image_path"`
	Images           StringSlice `gorm:"type:json;default:'[]'" json:"images"`
	Category         string      `gorm:"index" json:"category"`
	Difficulty       int         `json:"difficulty"`
	Tags             StringSlice `gorm:"type:json;default:'[]'" json:"tags"`
	Servings         int         `json:"servings"`
	PrepTimeMinutes  *int        `gorm:"column:prep_time_minutes" json:"prep_time_minutes"`
	CookTimeMinutes  *int        `gorm:"column:cook_time_minutes" json:"cook_time_minutes"`
	TotalTimeMinutes *int        `gorm:"column:total_time_minutes" json:"total_time_minutes"`

	// 关联关系
	Ingredients     []Ingredient     `gorm:"foreignKey:RecipeID;constraint:OnDelete:CASCADE" json:"ingredients"`
	Steps           []Step           `gorm:"foreignKey:RecipeID;constraint:OnDelete:CASCADE" json:"steps"`
	AdditionalNotes []AdditionalNote `gorm:"foreignKey:RecipeID;constraint:OnDelete:CASCADE" json:"additional_notes"`
}

func (Recipe) TableName() string {
	return "recipes"
}

// Ingredient 食材表
type Ingredient struct {
	ID           uint     `gorm:"primaryKey;autoIncrement" json:"id"`
	RecipeID     string   `gorm:"not null;index;column:recipe_id" json:"-"`
	Name         string   `gorm:"not null" json:"name"`
	Quantity     *float64 `json:"quantity"`
	Unit         *string  `json:"unit"`
	TextQuantity string   `gorm:"not null;column:text_quantity" json:"text_quantity"`
	Notes        *string  `json:"notes"`
}

func (Ingredient) TableName() string {
	return "ingredients"
}

// Step 步骤表
type Step struct {
	ID          uint   `gorm:"primaryKey;autoIncrement" json:"id"`
	RecipeID    string `gorm:"not null;index;column:recipe_id" json:"-"`
	Step        int    `gorm:"not null" json:"step"`
	Description string `gorm:"not null;type:text" json:"description"`
}

func (Step) TableName() string {
	return "steps"
}

// AdditionalNote 小贴士表
type AdditionalNote struct {
	ID       uint   `gorm:"primaryKey;autoIncrement" json:"id"`
	RecipeID string `gorm:"not null;index;column:recipe_id" json:"-"`
	Note     string `gorm:"not null;type:text" json:"note"`
}

func (AdditionalNote) TableName() string {
	return "additional_notes"
}
